Docs for bit pattern UDF interface

jklukas opened the pull request:
~~Most of the UDFs referenced here don't actually exist yet. The idea is to document bit pattern usage assuming these exist in order to gather feedback and update the design before putting the effort into implementation.~~

These UDFs now exist and are published.

wlach commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+# Working with Bit Patterns in Clients Last Seen
+
+An powerful feature of the `clients_last_seen` methodology is that we can capture
```
```suggestion
A powerful feature of the `clients_last_seen` methodology is that we can capture
```

wlach commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+# Working with Bit Patterns in Clients Last Seen
```
You may already have been thinking about this, but I think we need some kind of introductory paragraph(s) to motivate this content. Describing some concrete questions that might be answered by this material would be good.

wlach commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+For DAU, WAU, and MAU, all that matters is the rightmost bit that's set,
+telling us about the most recent activity seen for that user. We provide
+a special function for that use case:
+
```
```suggestion
```

wlach commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+later.
+
+When we were talking about backward-looking windows, our anchor date or "day 0"
+was always the `submission_date`. When we define forward-looking windows, however,
```
What does the `submission_date` actually means here? I assume you mean `submission_date` of clients_last_seen?

jklukas replied:
Yes, I'll try to make this more clear.

jmccrosky replied:
One could argue that we should not use the name `submission_date` at all in these tables, since this is far enough from raw pings that there is no reasonable definition of when anything was "submitted".  I haven't thought deeply about this though, so feel free to disagree :)

magorlick replied:
I prefer more specificity.  If these were all derived from submission dates that makes sense to me. If they are all different kinds of dates that were processed downstream to fit into a model of the truth we can use some other word.  I have a strong aversion to just using 'date' because it is a SQL function and I have to start putting quotes on it all the time.

wlach commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+we always choose an anchor date some time in the past. How we number the individual
+bits depends on what anchor date we choose.
+
+In GUD, we show a "Week 1 Retention" which considers a window of 14 days.
```
```suggestion
In [GUD](../tools/gud.md), we show a "Week 1 Retention" which considers a window of 14 days.
```

wlach commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+Let's consider the same user from before, grabbing the `days_seen_bits` value
+with `submission_date = 2020-01-28`. We'll still choose a window that _ends_
+on the submission date, but the anchor date (day 0) will now be 13 days earlier:
+
```
```suggestion
```

wlach commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,136 @@
+                                                    Week 0       └───────────────────┘
+                                                                        Week 1
+```
+
````
I think some kind of explanation of what we just determined by this calculation would be helpful here, instead of relying on the reader to fill in the blanks.

jklukas replied:
Will do.

wlach commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+(2020-01-28 in this case) and count backwards to "day -27" (2020-01-01).
+
+Let's suppose this client was only active on two days in the past month:
+2020-01-22 (day -6) and 2020-01-15 (day -13). That client's`days_seen_bits`
```
```suggestion
2020-01-22 (day -6) and 2020-01-15 (day -13). That client's `days_seen_bits`
```

jklukas commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,306 @@
+  submission_date = '2020-01-28'
+```
+
+### Retention using activity date
````
This is getting fairly speculative here. I've never actually run any queries using these activity date-based bit patterns, so they're unvetted. But I'm fairly confident that the kind of calculation described here is feasible.

magorlick replied:
Do we want to keep this or do we want to fairly firmly be in the submission_date camp?

jklukas replied:
I want to have some advanced examples. I've added more of a disclaimer here about how it's unofficial.

jmccrosky commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,375 @@
+
+For example, in [GUD](../tools/gud.md), we show a "1-Week Retention" which considers a window of 14 days.
+For each client active in "week 0" (days 0 through 6), we determine retention by
+checking if they were also active in "week 1" (days 0 through 13).
```
shouldn't this be "(days 7 through 13)"?

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+# Working with Bit Patterns in Clients Last Seen
+
+An powerful feature of the `clients_last_seen` methodology is that we can capture
+per-day usage information for the entire 28-day period that each row in the table
```
This reads a little bit confusing.  How can it be per day and for an entire 28 day period in one row? We are doing a rolling 28 day aggregation here that is tied to a day?  I thought it was fairly flexible to roll up info that occurred over the last 28 days in whatever way you want (could be less than 28 day roll up you just don't have access to things occurring outside of that window?  I always like to call out exactly what makes each row unique when I describe a table.

magorlick replied:
Also calling this out as a limitation of clients last seen (only 28 days) is useful.

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+per-day usage information for the entire 28-day period that each row in the table
+describes. We encode usage information in "bit patterns" where the physical
+type of the field is a BigQuery 64-bit integer, but logically the integer
+represents an array of bits, with each 1 indicating a day where the given clients
```
How is the array ordered?

jklukas replied:
I'm hoping this is addressed by the new "Understanding bit patterns" section, but let me know if it still feels like that needs to be addressed here.

magorlick replied:
Yes, this is clear below.

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+provide logic in views to insulate users from having to ever consider the
+raw bit patterns, but new investigations may require novel retention definitions.
+
+## Usage: Backward-looking windows
```
Don't we look forward in GUD?  For the record I prefer backward.  It raises less questions from folks that wonder where the recent data is.

jklukas replied:
WAU and MAU are backward-looking metrics where the value for 2020-01-28 depends on activity from 2020-01-01 to 2020-01-28.

Retention metrics are forward-looking where the value for 2020-01-28 depends on activity happening _after_ that date.

This is confusing and subtle for those who have to calculate the metrics, but I do think it's the most straight-forward approach from the consumer side. For a given date, you may want to compare 3-day retention, 1-week retention, and 2-week retention; this only makes sense if we align them based on the first day of the window.

jklukas replied:
I added a new "Forward-looking windows and backward-looking windows" section above this to hopefully capture the gotchas here.

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+DAU, WAU, and MAU.
+
+Let's imagine a single row from `clients_last_seen` with `submission_date = 2020-01-28`.
+The `days_seen_bits` field records usage for a single client over a period of 28 days
```
Any usage criteria baked in here? Does that limit this approach?

jklukas replied:
I will need to add some discussion of usage criteria, but yes usage criteria have to be fixed as part of the `clients_last_seen` where each usage criterion is its own bit pattern field. It's a known deficiency right now that we don't have a defined workflow for how DS can quickly test out new usage criterion definitions and propose them for inclusion. We need to do a full backfill to add a new usage criterion.

magorlick replied:
I like the discussion above.  I would call out the challenges to add new usage criteria as noted.

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,136 @@
+Let's make that representation a bit more visual:
+
+```
+    0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0  1  0  0  0  0  0  0
````
nice

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+won't count toward DAU for 2020-01-28, but the client does count towards both
+WAU and MAU.
+
+You may notice that the day 13 bit doesn't end up affecting these metrics.
```
It does affect MAU, but the activity on day 6 also affects MAU so you don't 'need' to see them on the 13th to count them in this specific case.

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,136 @@
+WAU and MAU.
+
+You may notice that the day 13 bit doesn't end up affecting these metrics.
+For DAU, WAU, and MAU, all that matters is the rightmost bit that's set,
```
You totally lost me here.  Imagine I have never studied bits before.  What do you mean by the rightmost bit? 0 on day 0 or 6 from the inscrutable 8256?

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,136 @@
+The `days_seen_bits` value is now `16512` and the user is indeed outside the WAU window:
+
+```
+SELECT udf.bits28_days_since_seen(16512)
````
This is still a mystery to me but maybe it is ok if folks don't understand how you go from the inscrutable number to 7 days since last seen and can just rely on the UDF

jklukas commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,375 @@
+```
+SELECT
+  submission_date,
+  udf.bits28_retention(days_seen_bits, submission_date).day_13.*
````
I'd appreciate thoughts on the interface for this `bits28_retention` function. I wanted to make it easy to use and in particular wanted to limit the number of parameters that you have to pass in. The idea of having this return a whole struct of the various retention metrics is a bit weird, but hopefully is convenient?

Is it clear from the description how to use this?

jklukas commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,375 @@
+Return an INT64 representing a range of bits from a source bit pattern.
+
+```
+bits28_range(bits INT64, offset_to_day_0 INT64, start_bit INT64, number_of_bits INT64)
````
This ` bits28_range` function is the trickiest and has the highest margin for error. It has a gaggle of params, all of which are ints, so ripe for getting the order wrong, etc.

This function also kind of does a lot. We could instead express this as two separate functions, one that trims off bits from the front and one that trims off bits from the end, something like:

```
-- Get bits 7 though 13.
bits28_take_first(bits28_take_last(bits, 14), 7)
```

jklukas replied:
I'm thinking that potentially a more simplistic function would actually be easier to reason about if we provide sufficient documentation on usage.

We could have `bits28_range(bits, start_index, number_of_bits)` where the index is a negative number that indicates a number of bits from the end of the pattern. So dealing with a 2-week window could look like:

```
SELECT
  DATE_SUB(submission_date, INTERVAL 13 DAY) AS metric_date,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 0, 1)) > 0 AS active_on_metric_date,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 0, 7)) > 0 AS active_in_week_0,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 1, 6)) > 0 AS active_in_week_0_after_metric_date,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 7, 7)) > 0 AS active_in_week_1,
```

hamilton commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,487 @@
+
+If you look at the `days_seen_bits` field in `telemetry.clients_last_seen`,
+you'll see seemingly random whole numbers, some as large as nine digits.
+How can we interpret these.
```
```suggestion
How should we interpret these?
```

fbertsch commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,487 @@
+
+```
+SELECT
+  -- Signature is bits28_range(offset_to_day_0, start_bit, number_of_bits)
````
This says 3 args, but below has 4 args.

jklukas replied:
This is an error right now.

jklukas commented:
These UDFs are all now implemented in https://github.com/mozilla/bigquery-etl/pull/948 and I've gone ahead and published them so that folks can start playing with them. I've also fixed various issues in the text here to make sure the examples match the published UDFs.

fbertsch commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,539 @@
+  telemetry.clients_last_seen
+
+/*
+| submission_date | metric_date | active_on_day_0 | active_in_week_0 | active_in_week_1 |
```
This seems to be missing `active_in_week_0_after_metric_date`. What is that used for?

jklukas replied:
I added an additional field to the full retention query example that uses this field.

fbertsch commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,539 @@
+
+The `bits28_range` function is powerful and flexible, but a bit cumbersome.
+We provide a `bits28_retention` convenience function that returns a nested
+structure with the standard retention definitions already calculated:
```
Is `standard retention definitions already calculated` the correct term here? I believe it should be `standard activity definitions already calculated`. We use these activity definitions to derive retention.

jklukas commented:
@felixlawrence - I'd be interested in your review here. Is this documentation sufficient that you'd feel confident being able to express your recent segmentation definitions using these UDFs? Does this interface feel more usable than existing UDFs?

felixlawrence commented on src/cookbooks/clients_last_seen_bits.md:
````diff
@@ -0,0 +1,553 @@
+Return an INT64 representing a range of bits from a source bit pattern.
+
+```
+bits28_range(bits INT64, offset INT64, n_bits INT64)
````
These methods seem to aim to remove the need for users to do bitwise operations explicitly. This is probably sensible: the learning curve is steep (though also short). Though these methods are harder to debug: I can't just print out the string representation of the mask to check that I'm doing what I expected.

It took me a lot of thinking to work out whether `offset` is with respect to the leftmost or rightmost bit, and the direction in which `n_bits` counts.

Given the rest of the docs, I would expect this to be backwards-looking: I would expect `n_bits` to count leftwards and an `offset` of `0` to represent the inclusion of the rightmost bit.

But from the examples it _seems_ that `n_bits` counts rightwards? So this function is most useful for the forwards-looking usecase? Do we expect the forwards-looking usecase to be more common? Should we have separate UDFs for forwards and backwards looking usecases?

As someone most familiar with Python, I (selfishly) want to think about this stuff in terms of slicing a flipped list, with arguments `start` and `stop`. So these examples would read `udf.bits28_range(18, 0, 6)`, `udf.bits28_range(18, 4, 6)` and `udf.bits28_range(18, 0, 4)` respectively.

jklukas replied:
> But from the examples it seems that n_bits counts rightwards? So this function is most useful for the forwards-looking usecase? Do we expect the forwards-looking usecase to be more common? Should we have separate UDFs for forwards and backwards looking usecases?

You are correct on all accounts. For backwards-looking use cases, `bits28_days_since_seen` is the only UDF needed. `bits28_range`, `bits28_active_in_range`, and `bits28_retention` are all intended for forward-looking windows, so they expect a negative offset to determine how many bits left to move before it starts reading, then `n_bits` measures how many bits to the right to move.

jklukas replied:
> As someone most familiar with Python, I (selfishly) want to think about this stuff in terms of slicing a flipped list, with arguments start and stop

What you describe is definitely more straight-forward in terms of how the parameters relate to the bit pattern. But that interface doesn't seem helpful for thinking about forward-looking windows where I think the right thing to do is to define a starting bit and then count rightward.

Does that sound reasonable? I can certainly add more to the documentation of this function to clarify some of the points you found surprising.

felixlawrence replied:
> For backwards-looking use cases, bits28_days_since_seen is the only UDF needed.

Not for my backwards-looking use cases! :sweat_smile:

Since the text above already discusses both forward and backward windows at length, it probably makes sense to specify which of these UDFs are counting forward and which are counting backward. At least here in the docs; optionally also in the naming of the functions, e.g. `bits28_forward_range()`.

felixlawrence replied:
> Does that sound reasonable?

Yes! As long as it's clear that they're for forward looking windows

jklukas replied:
Well, we could modify this function to allow negative numbers for `n_bits`. Negative `n_bits` would mean that we count left from the start offset (backward-looking windows) while positive `n_bits` would mean counting right (forward-looking windows).

felixlawrence replied:
For range specification/slicing, `(start, stop)` is an established format, as is `(start, length)` for positive lengths. I don't think I've seen `(start, length)` overloaded to become `(stop, negative_length)` when `length < 0`.

I would be a bit too scared to use it and will probably stick with masks (I'm OK with that - past the learning curve now) - I would worry about messing up the arithmetic and doing off-by-one miscalculations that take me to the 29th bit, yielding plausible but wrong results.

jklukas replied:
> I would worry about messing up the arithmetic and doing off-by-one miscalculations that take me to the 29th bit

We already have validation logic in this UDF such that we'll raise an error `'Reading %i bits from starting_offset %i exceeds end of bit pattern in call bits28_range(%i, %i, %i)'`. We would add additional checks if we enabled negative `n_bits` to at least catch the "29th bit" case.

> For range specification/slicing, (start, stop) is an established format, as is (start, length) for positive lengths. I don't think I've seen (start, length) overloaded to become (stop, negative_length) when length < 0.

I feel you here. I'm not sure I've ever seen an API that's quite analogous to what we have here. To be logically consistent for forward and backward-looking cases, I think we need to be consistent in thinking of offsets in terms of higher values as we move to the right. But maybe you're right that it would be easier to reason about if we had separate `bits28_range_forward` and `bits28_range_backward` or `bits28_range_right` and `bits28_range_left`.

jklukas commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,563 @@
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_active_in_range(days_seen_bits, -n + 0, 1) AS seen_on_day_0,
+    udf.bits28_active_in_range(days_seen_bits, -n + 1, n - 1) AS seen_after_day_0
```
```suggestion
    udf.bits28_active_in_range(days_seen_bits, -n + 1, 1) AS seen_on_day_0,
    udf.bits28_active_in_range(days_seen_bits, -n + 2, n - 1) AS seen_after_day_0
```

I think I made a mistake here.

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,689 @@
+
+Note that the desktop `clients_last_seen` table also has additional bit pattern
+fields corresponding to additional [usage criteria](../metrics/index.md),
+so other variants on MAU can be calculated like:
```
Do you want to put forward any comments on what it takes to get new usage critieria? Why these usage criteria?  My understanding is that this criteria is not flexible in production as is, instead we need to make a formal request for each additional usage criteria.

magorlick commented on src/cookbooks/clients_last_seen_bits.md:
```diff
@@ -0,0 +1,689 @@
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_active_in_range(days_seen_session_start_bits, offset_to_day_0 + 0, 1) AS seen_on_day_0,
```
Do we need to update this query as well?

    udf.bits28_active_in_range(days_seen_session_start_bits, offset_to_day_0+ 1, 1) AS seen_on_day_0,
    udf.bits28_active_in_range(days_seen_session_start_bits, offset_to_day_0 + 2, n - 1) AS seen_after_day_0

jklukas commented:
Looking for a final read-through at this point and a concrete approving review, then we'll merge this.
